README.md
# Dongtai-Base-Image
[中文版本(Chinese version)](README.ZH-CN.MD)

`dongtai-base-image` provides the basic environment for `DongTai` to run, including `MySql` and `Redis`

### Usage
- Build `MySql` service, `Redis` service using official image
- Build the `DongTai IAST` database from raw SQL files using custom `MySql` services and `Redis` services

### documentation
- [official doc](https://hxsecurity.github.io/DongTai-Doc/)
- [DongTai IAST WebSite](https://iast.huoxian.cn)


CHANGELOG.md
# Change log

## [1.2.0](https://github.com/HXSecurity/Dongtai-Base-Image/releases/tag/v1.2.0)-2021-12-31

* Features
     * Add component LICENSE display
* Improve
     * Increase index to improve query speed
     * Synchronized the latest agent hook strategy
     * Synchronized mobile phone number leakage rules @SpenserCai

## [1.1.4](https://github.com/HXSecurity/Dongtai-Base-Image/releases/tag/v1.1.4)-2021-12-18

* Improve
     * Add python hook rules
     * Increase index to improve query time
     
## [1.1.3](https://github.com/HXSecurity/Dongtai-Base-Image/releases/tag/v1.1.3)-2021-12-03

* Function
   * Increase the vulnerability active verification switch (including global and project level)
* Improve
   * Add strategy 
   * Add sensitive_info rule


test/test_strategy.py
from configparser import ConfigParser
import os
import pymysql
pymysql.install_as_MySQLdb()
import MySQLdb
import sys
config = ConfigParser()
status = config.read('conf/config.ini')
if len(status) == 0:
    print("config file not exist. stop running")
    sys.exit(0)


DBCONFIG = {
    'user': config.get("mysql", 'user'),
    'db': config.get("mysql", 'name'),
    'passwd': config.get("mysql", 'password'),
    'host': config.get("mysql", 'host'),
    'port': int(config.get("mysql", 'port')),
}
TARGETDBCONFIG = {
    'user': config.get("prodmysql", 'user'),
    'db': config.get("prodmysql", 'name'),
    'passwd': config.get("prodmysql", 'password'),
    'host': config.get("prodmysql", 'host'),
    'port': int(config.get("prodmysql", 'port')),
}
#class TestDB:
#    db = MySQLdb.connect(**DBCONFIG, use_unicode=True, charset="utf8mb4",cursorclass=pymysql.cursors.DictCursor)
class TargetDB:
    db = MySQLdb.connect(**TARGETDBCONFIG, use_unicode=True, charset="utf8mb4",cursorclass=pymysql.cursors.DictCursor)
def test_hook_strategy():
    cursor = TestDB.db.cursor()
    cursor.execute("SELECT * FROM iast_hook_type;")
    data = cursor.fetchall()
    test_hook_type= {}
    for item in data:
        test_hook_type[item['name']]=item
    cursor = TargetDB.db.cursor()
    cursor.execute("SELECT * FROM iast_hook_type;")
    data = cursor.fetchall()
    target_hook_type = {}
    for item in data:
        target_hook_type[item['name']] = item
    for key, item in target_hook_type.items():
        if key not in test_hook_type.keys():
            print(item)


utils/test_strategy.py
from configparser import ConfigParser
import os
import pymysql
pymysql.install_as_MySQLdb()
import MySQLdb
import sys
config = ConfigParser()
status = config.read('conf/config.ini')
if len(status) == 0:
    print("config file not exist. stop running")
    sys.exit(0)
DBCONFIG = {
    'user': config.get("mysql", 'user'),
    'db': config.get("mysql", 'name'),
    'passwd': config.get("mysql", 'password'),
    'host': config.get("mysql", 'host'),
    'port': int(config.get("mysql", 'port')),
}
TARGETDBCONFIG = {
    'user': config.get("prodmysql", 'user'),
    'db': config.get("prodmysql", 'name'),
    'passwd': config.get("prodmysql", 'password'),
    'host': config.get("prodmysql", 'host'),
    'port': int(config.get("prodmysql", 'port')),
}
class TestDB:
    db = MySQLdb.connect(**DBCONFIG, use_unicode=True, charset="utf8mb4",cursorclass=pymysql.cursors.DictCursor)
class TargetDB:
    db = MySQLdb.connect(**TARGETDBCONFIG, use_unicode=True, charset="utf8mb4",cursorclass=pymysql.cursors.DictCursor)


def test_hook_strategy():
    cursor = TestDB.db.cursor()
    cursor.execute("SELECT * FROM iast_hook_type;")
    data = cursor.fetchall()
    test_hook_type= {}
    for item in data:
        test_hook_type[item['name']]=item
    cursor = TargetDB.db.cursor()
    cursor.execute("SELECT * FROM iast_hook_type;")
    data = cursor.fetchall()
    target_hook_type = {}
    for item in data:
        target_hook_type[item['name']] = item
    for key, item in target_hook_type.items():
        if key not in test_hook_type.keys():
            print(item)


test/hook_strategy_export.py
######################################################################
# @author      : bidaya0 (bidaya0@$HOSTNAME)
# @file        : hook_strategy_export
# @created     : 星期一 2月 07, 2022 16:29:01 CST
#
# @description :
######################################################################

from test_strategy import TargetDB

cursor = TargetDB.db.cursor()
hook_type_query_sql = '''
select * from iast_hook_type iht 
where id in (select hooktype_id from iast_hook_strategy_type ihst 
where  hooktype_id in (
select id from iast_hook_type iht where  language_id  = 1 AND enable in (0,1) and `type` in (1,2,3,4) )
and hookstrategy_id  in (select id from iast_hook_strategy ihs where created_by = 1 and enable in (0,1)));
'''

hook_type_sql = '''
INSERT IGNORE INTO `iast_hook_type` (`type`, `name`, `value`, `create_time`
	, `update_time`, `created_by`, `enable`, `name_en`, `name_zh`
	, `language_id`)
SELECT %s, %s, %s, %s
	, %s, %s, %s, %s, %s
	, %s FROM DUAL WHERE NOT EXISTS (SELECT `id` FROM iast_hook_type WHERE 
`type`=%s AND `name`= %s
AND  value = %s AND create_time= %s AND
update_time = %s  AND `created_by`= %s
AND enable = %s AND name_en = %s AND name_zh = %s AND language_id = %s LIMIT 1);
SET @HOOK_TYPE_ID = (SELECT `id` FROM iast_hook_type WHERE 
`type`=%s AND `name`= %s
AND  value = %s AND create_time = %s AND
update_time = %s  AND `created_by` = %s
AND enable = %s AND name_en = %s AND name_zh = %s AND language_id = %s LIMIT 1);
'''

hook_strategy_query_sql = '''
select * from iast_hook_strategy ihs2 
where id in (select hookstrategy_id from iast_hook_strategy_type ihst 
where  hooktype_id in (
select id from iast_hook_type iht where  language_id  = 1 AND enable in (0,1) and `type` in (1,2,3,4))
and hookstrategy_id  in (select id from iast_hook_strategy ihs where created_by = 1 and enable in (0,1)));
'''

hook_strategy_sql = '''
INSERT IGNORE INTO iast_hook_strategy
(value, source, target, inherit, track, create_time, update_time, created_by, enable)
SELECT %s, %s, %s, %s, %s, %s, %s, %s, %s FROM DUAL 
WHERE NOT EXISTS (SELECT `id` FROM iast_hook_strategy WHERE 
`value` = %s AND `source` = %s AND `target` = %s AND `inherit` = %s AND `track` = %s AND `create_time` = %s AND `update_time` = %s AND `created_by` = %s AND `enable` = %s LIMIT 1);
SET @IAST_HOOK_STRATEGY_ID = (SELECT `id` FROM iast_hook_strategy WHERE 
`value` = %s AND `source` = %s AND `target` = %s AND `inherit` = %s AND `track` = %s AND `create_time` = %s AND `update_time` = %s AND `created_by` = %s AND `enable` = %s LIMIT 1);
'''

hook_strategy_type_query_sql = '''
select * from iast_hook_strategy_type ihst
where  hooktype_id in (
select id from iast_hook_type iht where  language_id  = 1 AND enable in (0,1) and `type` in (1,2,3,4))
and hookstrategy_id  in (select id from iast_hook_strategy ihs where created_by = 1 and enable in (0,1));
'''

finsql = hook_strategy_sql.format(
    **{
        i: '%s'
        for i in [
            'value', 'source', 'target', 'inherit', 'track', 'create_time',
            'update_time', 'created_by', 'enable'
        ]
    })


cursor.execute(hook_type_query_sql)


def escape_null_equal(sql):
    return sql.replace('= NULL', 'IS NULL')


hook_type_dict = {}
for data in cursor.fetchall():
    final_sql = cursor.mogrify(hook_type_sql, [
        data['type'],
        data['name'],
        data['value'],
        data['create_time'],
        data['update_time'],
        data['created_by'],
        data['enable'],
        data['name_en'],
        data['name_zh'],
        data['language_id'],
    ] * 3)
    hook_type_dict[int(data['id'])] = escape_null_equal(final_sql)




cursor.execute(hook_strategy_query_sql)


def escape_null_equal(sql):
    return sql.replace('= NULL', 'IS NULL')


hook_strategy_dict = {}
for data in cursor.fetchall():
    final_sql = cursor.mogrify(hook_strategy_sql, [
        data[i] for i in [
            'value', 'source', 'target', 'inherit', 'track', 'create_time',
            'update_time', 'created_by', 'enable'
        ]
        ] * 3)
    hook_strategy_dict[int(data['id'])] = escape_null_equal(final_sql)


cursor.execute(hook_strategy_type_query_sql)
for data in cursor.fetchall():
    print(hook_type_dict[data['hooktype_id']])
    print(hook_strategy_dict[data['hookstrategy_id']])
    print(
        'INSERT INTO iast_hook_strategy_type (hookstrategy_id, hooktype_id) VALUES (@IAST_HOOK_STRATEGY_ID, @HOOK_TYPE_ID);'
    )

#print(hook_strategy_dict)


utils/hook_rull_export_v3.py
######################################################################
# @author      : bidaya0 (bidaya0@$HOSTNAME)
# @file        : hook_type_rull_sql
# @created     : 星期五 10月 22, 2021 21:41:12 CST
#
# @description :
######################################################################

with open('complitejava.sql', 'r') as fp:
    sql = fp.readlines()
hook_type_dict = {}
hook_strategy_pair = []
hook_strategy_dict = {}
	
for i in sql:
    if i.startswith(
            'INSERT IGNORE INTO iast_hook_type (id, `type`, name, value, create_time, update_time, created_by, enable, name_en, name_zh, language_id, strategy_id) VALUES('
    ):
        a = i.replace(
            'INSERT IGNORE INTO iast_hook_type (id, `type`, name, value, create_time, update_time, created_by, enable, name_en, name_zh, language_id, strategy_id) VALUES(',
            '').replace(');\n', '')
        id_, type_, name, value, create_time, update_time, created_by, enable, name_en, name_zh, language_id, strategy_id = res = a.split(
            ',')
        print(res)
        hook_type_dict[int(id_)] ='''INSERT IGNORE INTO `iast_hook_type` (`type`, `name`, `value`, `create_time`
	, `update_time`, `created_by`, `enable`, `name_en`, `name_zh`
	, `language_id`)
SELECT {type_}, {name}, {value}, {create_time}
	, {update_time}, {created_by}, {enable}, {name_en}, {name_zh}
	, {language_id} FROM DUAL WHERE NOT EXISTS (SELECT `id` FROM iast_hook_type WHERE 
`type`={type_} AND `name`= {name}
AND  value = {value} AND
update_time={update_time} AND create_time={create_time} AND `created_by`={created_by}
AND enable = {enable} AND name_en = {name_en}AND name_zh = {name_zh} AND language_id = {language_id} LIMIT 1);
SET @HOOK_TYPE_ID = (SELECT `id` FROM iast_hook_type WHERE 
`type`={type_} AND `name`= {name}
AND  value = {value} AND
update_time={update_time} AND create_time={create_time} AND `created_by`={created_by}
AND enable = {enable} AND name_en = {name_en}AND name_zh = {name_zh} AND language_id = {language_id} LIMIT 1);

'''.format(
                type_=type_,
                name=name,
                value=value,
                create_time=create_time,
                update_time=update_time,
                created_by=created_by,
                enable=enable,
                name_en=name_en,
                name_zh=name_zh,
                language_id=language_id)
    elif i.startswith(
            'INSERT IGNORE INTO iast_hook_strategy (id, value, source, target, inherit, track, create_time, update_time, created_by, enable) VALUES('
    ):
        a = i.replace(
            'INSERT IGNORE INTO iast_hook_strategy (id, value, source, target, inherit, track, create_time, update_time, created_by, enable) VALUES(',
            '').replace(');', '')
        print(a.split(','))
        id_, value, source, target, inherit, track, create_time, update_time, created_by, enable = res = a.split(
            ',')
        print(res)
        hook_strategy_dict[int(id_)] = '''INSERT IGNORE INTO iast_hook_strategy
(value, source, target, inherit, track, create_time, update_time, created_by, enable)
SELECT {value}, {source}, {target}, {inherit}, {track}, {create_time}, {update_time}, {created_by}, {enable} FROM DUAL 
WHERE NOT EXISTS (SELECT `id` FROM iast_hook_strategy WHERE 
`value`={value} AND `source`={source} AND `target`={target} AND `inherit`={inherit} AND `track`={track} AND `create_time`= {create_time} AND `update_time`= {update_time} AND `created_by`={created_by} AND `enable` = {enable} LIMIT 1);
SET @IAST_HOOK_STRATEGY_ID = (SELECT `id` FROM iast_hook_strategy WHERE 
`value`={value} AND `source`={source} AND `target`={target} AND `inherit`={inherit} AND `track`={track} AND `create_time`= {create_time} AND `update_time`= {update_time} AND `created_by`={created_by} AND `enable` = {enable} LIMIT 1);
'''.format(value=value,
           source=source,
           target=target,
           inherit=inherit,
           track=track,
           create_time=create_time,
           update_time=update_time,
           created_by=created_by,
           enable=enable)
    elif i.startswith(
            'INSERT IGNORE INTO iast_hook_strategy_type (id, hookstrategy_id, hooktype_id) VALUES('
    ):
        a = i.replace(
            'INSERT IGNORE INTO iast_hook_strategy_type (id, hookstrategy_id, hooktype_id) VALUES(',
            '').replace(');', '')
        id_, hookstrategy_id, hooktype_id = res = a.split(',')
        hook_strategy_pair.append([int(hookstrategy_id), int(hooktype_id)])


for k, v in hook_strategy_pair:
    res = hook_type_dict[int(v)]
    print(hook_strategy_dict[int(k)])
    print(
        'INSERT IGNORE INTO iast_hook_strategy_type (hookstrategy_id, hooktype_id) VALUES (@IAST_HOOK_STRATEGY_ID, @HOOK_TYPE_ID);'
    )


utils/hook_strategy_export.py
######################################################################
# @author      : bidaya0 (bidaya0@$HOSTNAME)
# @file        : hook_strategy_export
# @created     : 星期一 2月 07, 2022 16:29:01 CST
#
# @description :
######################################################################

from test_strategy import TestDB,TargetDB

cursor = TargetDB.db.cursor()
hook_type_query_sql = '''
select * from iast_hook_type iht 
where id in (select hooktype_id from iast_hook_strategy_type ihst 
where  hooktype_id in (
select id from iast_hook_type iht where  language_id IN (1,2,3,4) AND enable in (0,1) and `type` in (1,2,3,4) )
and hookstrategy_id  in (select id from iast_hook_strategy ihs where created_by IN (1,1207) and enable in (0,1)));
'''

hook_type_sql = '''
INSERT IGNORE INTO `iast_hook_type` (`type`, `name`, `value`
	, `created_by`, `enable`, `name_en`, `name_zh`
	, `language_id`)
SELECT %s, %s, %s, %s
	, %s, %s, %s 
	, %s FROM DUAL WHERE NOT EXISTS (SELECT `id` FROM iast_hook_type WHERE 
`type`=%s AND `name`= %s
AND  value = %s AND  `created_by`= %s
AND enable = %s AND name_en = %s AND name_zh = %s AND language_id = %s LIMIT 1);
SET @HOOK_TYPE_ID = (SELECT `id` FROM iast_hook_type WHERE 
`type`=%s AND `name`= %s
AND  value = %s  AND `created_by` = %s
AND enable = %s AND name_en = %s AND name_zh = %s AND language_id = %s LIMIT 1);
'''

hook_strategy_query_sql = '''
select * from iast_hook_strategy ihs2 
where id in (select hookstrategy_id from iast_hook_strategy_type ihst 
where  hooktype_id in (
select id from iast_hook_type iht where  language_id  IN (1,2,3,4) AND enable in (0,1) and `type` in (1,2,3,4))
and hookstrategy_id  in (select id from iast_hook_strategy ihs where created_by IN (1,1207) and enable in (0,1)));
'''

hook_strategy_sql = '''
INSERT IGNORE INTO iast_hook_strategy
(value, source, target, inherit, track, created_by, enable)
SELECT %s, %s, %s, %s, %s, %s, %s FROM DUAL 
WHERE NOT EXISTS (SELECT `id` FROM iast_hook_strategy WHERE 
`value` = %s AND `source` = %s AND `target` = %s AND `inherit` = %s AND `track` = %s AND  `created_by` = %s AND `enable` = %s LIMIT 1);
SET @IAST_HOOK_STRATEGY_ID = (SELECT `id` FROM iast_hook_strategy WHERE 
`value` = %s AND `source` = %s AND `target` = %s AND `inherit` = %s AND `track` = %s AND  `created_by` = %s AND `enable` = %s LIMIT 1);
'''

hook_strategy_type_query_sql = '''
select * from iast_hook_strategy_type ihst
where  hooktype_id in (
select id from iast_hook_type iht where  language_id  IN (1,2,3,4) AND enable in (0,1) and `type` in (1,2,3,4))
and hookstrategy_id  in (select id from iast_hook_strategy ihs where created_by IN (1,1207) and enable in (0,1));
'''

finsql = hook_strategy_sql.format(
    **{
        i: '%s'
        for i in [
            'value', 'source', 'target', 'inherit', 'track', 'created_by', 'enable'
        ]
    })


cursor.execute(hook_type_query_sql)


def escape_null_equal(sql):
    return sql.replace('= NULL', 'IS NULL')


hook_type_dict = {}
for data in cursor.fetchall():
    final_sql = cursor.mogrify(hook_type_sql, [
        data['type'],
        data['name'],
        data['value'],
        1,
        data['enable'],
        data['name_en'],
        data['name_zh'],
        data['language_id'],
    ] * 3)
    hook_type_dict[int(data['id'])] = escape_null_equal(final_sql)




cursor.execute(hook_strategy_query_sql)


def escape_null_equal(sql):
    return sql.replace('= NULL', 'IS NULL')


hook_strategy_dict = {}
for data in cursor.fetchall():
    data['created_by'] = 1
    final_sql = cursor.mogrify(hook_strategy_sql, [
        data[i] for i in [
            'value', 'source', 'target', 'inherit', 'track', 'created_by',
            'enable'
        ]
    ] * 3)
    hook_strategy_dict[int(data['id'])] = escape_null_equal(final_sql)

cursor.execute(hook_strategy_type_query_sql)
for data in cursor.fetchall():
    print(hook_type_dict[data['hooktype_id']])
    print(hook_strategy_dict[data['hookstrategy_id']])
    print(
        'INSERT IGNORE INTO iast_hook_strategy_type (hookstrategy_id, hooktype_id) VALUES (@IAST_HOOK_STRATEGY_ID, @HOOK_TYPE_ID);'
    )

hook_strategy_value_sql = '''
SELECT value FROM iast_hook_strategy ihs ;
'''

DELETE_EXIST_STRATEGY_SQL = '''
DELETE FROM iast_hook_strategy WHERE value IN ({});
'''
testcursor = TestDB.db.cursor()
testcursor.execute(hook_strategy_value_sql)
result = testcursor.fetchall()
values = map(lambda x: x['value'], result)
res = DELETE_EXIST_STRATEGY_SQL.format(','.join(
    cursor._escape_args(list(values), cursor._get_db())))
print(res)
